### PR TITLE
[XCR-13] Add Roles and Permissions

### PR DESCRIPTION
**Pull request by thongphp**

### CODE REVIEW

File: app/Core/EventSourcing/Listeners/RecordedEventSubscriber.php [hunk 1]
```diff
@@ -3,18 +3,26 @@
...
-use App\Core\EventSourcing\RecordedEvent;
+use App\Core\EventSourcing\RecordedEventInterface;
 
 class RecordedEventSubscriber
 {
     private EventManager $eventManager;
 
+    /**
+     * @param \App\Core\EventSourcing\EventManager $eventManager
+     */
```
Comment (jooservices) [inline app/Core/EventSourcing/Listeners/RecordedEventSubscriber.php:14]: For next time docblock is not required

Comment (jooservices) [reply app/Core/EventSourcing/Listeners/RecordedEventSubscriber.php:14]: We only need it for
- Specific case when will we need explain something
- Type hint is more important

File: app/Models/Permission.php [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+ * @method static \Illuminate\Database\Eloquent\Builder|Permission whereDescription($value)
+ * @method static \Illuminate\Database\Eloquent\Builder|Permission whereDisplayName($value)
+ * @method static \Illuminate\Database\Eloquent\Builder|Permission whereId($value)
+ * @method static \Illuminate\Database\Eloquent\Builder|Permission whereName($value)
+ * @method static \Illuminate\Database\Eloquent\Builder|Permission whereUpdatedAt($value)
+ * @mixin \Eloquent
+ */
+class Permission extends LaratrustPermission
+{
+    public $guarded = [];
```
Comment (jooservices) [inline app/Models/Permission.php:31]: Why use empty guarded ???
We need

- $fillable
- $casts

File: database/migrations/2021_03_29_184820_add_columns_for_events_table.php [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up(): void
+    {
+        Schema::table('events', function (Blueprint $table) {
+            $table->string('ip_address')->after('data');
```
Comment (jooservices) [inline database/migrations/2021_03_29_184820_add_columns_for_events_table.php:17]: No ! Use integer for storing IP

File: tests/Unit/User/UserRolePermissionTest.php [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        $adminRole = Role::create(['name' => 'admin']);
+        $userRole = Role::create(['name' => 'user']);
+
+        $adminRole->attachPermissions([$profileView, $profileEdit, $createUser, $deleteUser]);
+        $userRole->attachPermissions([$profileView, $profileEdit]);
+
+        /** @var User $admin */
+        $admin = User::Factory()->create();
+        $admin->attachRole($adminRole);
+        self::assertTrue($admin->isAbleTo(['profile-view', 'profile-edit', 'user-create', 'user-delete'], null, true));
```
Comment (jooservices) [inline tests/Unit/User/UserRolePermissionTest.php:28]: $this->assertTrue

File: tests/Unit/User/UserRolePermissionTest.php [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+namespace Tests\Unit\User;
+
+use App\Models\Permission;
+use App\Models\Role;
+use App\Models\User;
+use Tests\TestCase;
+
+class UserRolePermissionTest extends TestCase
+{
+    public function testRoleAndPermissions(): void
```
Comment (jooservices) [inline tests/Unit/User/UserRolePermissionTest.php:12]: Consider use dataSet

File: app/Models/Permission.php [hunk 6]
```diff
@@ -0,0 +1,32 @@
+<?php
+
+namespace App\Models;
+
+use Laratrust\Models\LaratrustPermission;
+
+/**
+ * App\Models\Permission
+ *
+ * @property int $id
```
Comment (jooservices) [inline app/Models/Permission.php:10]: These are required ! agree

**Review by jooservices — CHANGES_REQUESTED** [review]

**Comment by jooservices** [comment]
I have fixed build.yaml
Please rebase to develop and wait for UT are passed

File: app/Core/EventSourcing/Listeners/RecordedEventSubscriber.php (hunk 1 above)
Comment (thongphp) [reply app/Core/EventSourcing/Listeners/RecordedEventSubscriber.php:14]: oki

File: database/migrations/2021_03_29_184820_add_columns_for_events_table.php (hunk 3 above)
Comment (thongphp) [reply database/migrations/2021_03_29_184820_add_columns_for_events_table.php:17]: Are you sure? IP address is can not store in Integer in MySQL

File: tests/Unit/User/UserRolePermissionTest.php (hunk 5 above)
Comment (thongphp) [reply tests/Unit/User/UserRolePermissionTest.php:12]: We don't need for this case. Later on for more complex tests, we would use dataSet

File: database/migrations/2021_03_29_184820_add_columns_for_events_table.php [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up(): void
+    {
+        Schema::table('events', function (Blueprint $table) {
+            $table->integer('ip_address')->after('data');
```
Comment (jooservices) [inline database/migrations/2021_03_29_184820_add_columns_for_events_table.php:17]: `$table->bigInteger('ip_address')->nullable()->after('data')->index();`

**Review by jooservices — CHANGES_REQUESTED** [review]

File: tests/Unit/User/UserRolePermissionTest.php (hunk 5 above)
Comment (jooservices) [reply tests/Unit/User/UserRolePermissionTest.php:12]: Nop use dataSet for this case to make it cleaner

**Review by jooservices — CHANGES_REQUESTED** [review]

**Pull request closed without merging**